### PR TITLE
SNES/SFC file format, slight fix for analyzer

### PR DESCRIPTION
**Pull request by usrshare**
This pull requests adds code that recognizes the SNES/SFC ROM file format (should work with both headered and headerless files) that adds basic RAM and ROM sections to the file.

In addition, the analyzer for SNES code now interprets JMP and JSR operations as a jump within the current bank (e.g. at 0x031337, a JMP 0xC0DE will jump to 0x03C0DE).

(note: i tried running uncrustify with that config in the doc/ directory to make indentation look better, hopefully this works)

### CODE REVIEW

File: libr/bin/p/bin_sfc.c [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+
+static bool check_bytes(const ut8 *buf, ut64 length) {
+
+	const ut8* buf_hdr = buf;
+	ut16 cksum1, cksum2;
+
+	if ((length & 0x8000) == 0x200) buf_hdr += 0x200;
+
+	//determine if ROM is headered, and add a 0x200 gap if so. 
+	cksum1 = r_read_le16(buf_hdr + 0x7FDC);
```
Comment (Maijin) [inline libr/bin/p/bin_sfc.c]: Couldn't 0x7FDC and 0x7FDE be added as a define instead of magic number?

File: libr/bin/p/bin_sfc.c [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+#include <r_lib.h>
+#include "sfc/sfc_specs.h"
+#include <r_endian.h>
+
+static bool check_bytes(const ut8 *buf, ut64 length) {
+
+	const ut8* buf_hdr = buf;
+	ut16 cksum1, cksum2;
+
+	if ((length & 0x8000) == 0x200) buf_hdr += 0x200;
```
Comment (Maijin) [inline libr/bin/p/bin_sfc.c]: Couldn't 0x8000, 0x200 be added as define ?

File: libr/bin/p/bin_sfc.c [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+	ptr->ordinal = 0;
+	r_list_append (ret, ptr);
+}
+
+static RList* symbols(RBinFile *arch) {
+	RList *ret = NULL;
+	if (!(ret = r_list_new ()))
+		return NULL;
+	ret->free = free;
+	//addsym (ret, "NMI_VECTOR_START_ADDRESS", NMI_VECTOR_START_ADDRESS,2);
```
Comment (Maijin) [inline libr/bin/p/bin_sfc.c]: No symbol for SNES? If yes, please remove the NES comment those are irrelevant.

File: libr/bin/format/sfc/sfc_specs.h [hunk 4]
```diff
@@ -0,0 +1,58 @@
+/* radare - LGPL - 2015 - maijin */
```
Comment (Maijin) [inline libr/bin/format/sfc/sfc_specs.h]: Change with your name and date

File: libr/bin/p/bin_sfc.c [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+	return ret;
+}
+
+static RList* entries(RBinFile *arch) { //Should be 3 offsets pointed by NMI, RESET, IRQ after mapping && default = 1st CHR
+	RList *ret;
+	if (!(ret = r_list_new ())) {
+		return NULL;
+	}
+	/*
+	RBinAddr *ptr = NULL;
```
Comment (Maijin) [inline libr/bin/p/bin_sfc.c]: Remove the comments from NES

File: libr/bin/p/bin_sfc.c [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+			eprintf ("Cannot determine if this is a LoROM or HiROM file\n");
+			return NULL;
+		}
+		is_hirom = 1;
+	}
+	
+	if (!(ret = r_list_new ()))
+		return NULL;
+
+
```
Comment (Maijin) [inline libr/bin/p/bin_sfc.c]: Remove whitespace, use r2-indent tool

File: libr/bin/p/bin_sfc.c [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+		}
+	
+		if ( (sfchdr.comp_check != (ut16)~(sfchdr.checksum)) || ((sfchdr.rom_setup & 0x1) != 1) ) {
+
+			eprintf ("Cannot determine if this is a LoROM or HiROM file\n");
+			return NULL;
+		}
+	}
+	
+	if (!(ret = R_NEW0 (RBinInfo)))
```
Comment (radare) [inline libr/bin/p/bin_sfc.c]: use r2-indent

**Comment by radare** [comment]
Add it to the plugins.def.cfg and rebase

**Comment by usrshare** [comment]
Rebase? How exactly should I do this?

**Comment by Maijin** [comment]
https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

**Comment by usrshare** [comment]
Okay, now that looks even more wrong than before. Maybe I'll just scrap everything and do it from scratch.

**Comment by Maijin** [comment]
Never merge master :P

**Comment by Maijin** [comment]
@usrshare are you really going your work die completely ?

**Comment by radare** [comment]
I have rebased fixed and merged your PR, after it i did some fixes because there was a bunch of bugs, but now it should be ready. please test :) and thanks for the fish

**Comment by usrshare** [comment]
Dunno. I found that adding the virtual addressing sections manually is easier and more reliable. Plus, the entire algorithm made lots of assumptions, which I felt could lead to mistakes later on.

**Comment by radare** [comment]
Feel free to send a pr updating your parser. I fixed a bunch of null derefs and oobreads. Will be good to have some tests for this, do you know any opensource snes rom to use in yhe testsuite?

**Pull request closed without merging**